### DISCUSSION TITLE
Visual Studio Project files missing

### DISCUSSION THREAD

**Discussion by ganeshkamath89** [curl/curl#6427] [Q&A] [answered]
Curl solution does not have Visual Studio project in the location:
https://github.com/curl/curl/tree/master/projects/Windows/VC15/src

I looked for vcxproj file in all the folders within:
https://github.com/curl/curl/tree/master/projects/Windows/

Do I need to run some script to generate the vcxproj file?

**Comment by jay** [comment, marked as answer]
From the [README](https://github.com/curl/curl/blob/master/projects/README):
~~~
   If you are using the source code from the git repository, rather than a
   release archive or nightly build, you will need to generate the project
   files. Please run "generate -help" for usage details.
~~~

For example generate vc15

**Comment by ganeshkamath89** [comment]
Hi @jay. Thank you very much. I have generated the project file and using it successfully with the command "generate vc15".

I have 1 suggestion.

When I had tried "generate" without any parameters, instead of displaying the available options, the script kept on generating for all vc options. I think it goes into an infinite loop, because I manually closed that script after waiting for 30 minutes. I think it is looping because it kept displaying the same set of options after some time back to back.

**Reply by jay** [reply]
It should generate to vc15
~~~
X:\j\curl\curl\projects>generate

Generating prerequisite files
* X:\j\curl\curl\Makefile
* X:\j\curl\curl\src\tool_hugehelp.c

Generating VC6 project files
* X:\j\curl\curl\projects\Windows\VC6\src\curl.dsp
* X:\j\curl\curl\projects\Windows\VC6\lib\libcurl.dsp

Generating VC7 project files
* X:\j\curl\curl\projects\Windows\VC7\src\curl.vcproj
* X:\j\curl\curl\projects\Windows\VC7\lib\libcurl.vcproj

Generating VC7.1 project files
* X:\j\curl\curl\projects\Windows\VC7.1\src\curl.vcproj
* X:\j\curl\curl\projects\Windows\VC7.1\lib\libcurl.vcproj

Generating VC8 project files
* X:\j\curl\curl\projects\Windows\VC8\src\curl.vcproj
* X:\j\curl\curl\projects\Windows\VC8\lib\libcurl.vcproj

Generating VC9 project files
* X:\j\curl\curl\projects\Windows\VC9\src\curl.vcproj
* X:\j\curl\curl\projects\Windows\VC9\lib\libcurl.vcproj

Generating VC10 project files
* X:\j\curl\curl\projects\Windows\VC10\src\curl.vcxproj
* X:\j\curl\curl\projects\Windows\VC10\lib\libcurl.vcxproj

Generating VC11 project files
* X:\j\curl\curl\projects\Windows\VC11\src\curl.vcxproj
* X:\j\curl\curl\projects\Windows\VC11\lib\libcurl.vcxproj

Generating VC12 project files
* X:\j\curl\curl\projects\Windows\VC12\src\curl.vcxproj
* X:\j\curl\curl\projects\Windows\VC12\lib\libcurl.vcxproj

Generating VC14 project files
* X:\j\curl\curl\projects\Windows\VC14\src\curl.vcxproj
* X:\j\curl\curl\projects\Windows\VC14\lib\libcurl.vcxproj

Generating VC15 project files
* X:\j\curl\curl\projects\Windows\VC15\src\curl.vcxproj
* X:\j\curl\curl\projects\Windows\VC15\lib\libcurl.vcxproj
~~~

**Comment by ganeshkamath89** [comment]
Hi @jay 

I ran the generate command again. I am able to reproduce the behaviour:

~~~
[MyPath]>generate.bat

Generating prerequisite files
* [MyPath]\curl-curl-7_74_0\Makefile
* [MyPath]\curl-curl-7_74_0\src\tool_hugehelp.c

Warning: The curl manual could not be integrated in the source. This means when
you build curl the manual will not be available (curl --man). Integration of
the manual is not required and a summary of the options will still be available
(curl --help). To integrate the manual your PATH is required to have
groff/nroff, perl and optionally gzip for compression.

Generating VC6 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC6\src\curl.dsp
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC6\lib\libcurl.dsp

Generating VC7 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC7\src\curl.vcproj
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC7\lib\libcurl.vcproj

Generating VC7.1 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC7.1\src\curl.vcproj
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC7.1\lib\libcurl.vcproj

Generating VC8 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC11\lib\libcurl.vcxproj
Removing VC11 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC11\src\curl.vcxproj
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC11\lib\libcurl.vcxproj

Generating VC12 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC12\src\curl.vcxproj
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC12\lib\libcurl.vcxproj

Generating VC14 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC14\src\curl.vcxproj
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC14\lib\libcurl.vcxproj

Generating VC15 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC15\src\curl.vcxproj
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC15\lib\libcurl.vcxproj
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC11\lib\libcurl.vcxproj
Removing VC11 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC11\src\curl.vcxproj
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC11\lib\libcurl.vcxproj

Generating VC12 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC12\src\curl.vcxproj
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC12\lib\libcurl.vcxproj

Generating VC14 project files
* [MyPath]\curl-curl-7_74_0\projects\Windows\VC14\src\curl.vcxproj
~~~

**Reply by jay** [reply]
So you are saying it hangs at VC14? It should only take a minute to generate all the files. Maybe more if you have an anti-virus client or are generating on a network drive.

**Comment by ganeshkamath89** [comment]
Hi @jay . When I run the batch file on my local system, it continues to run the same set of commands over and over again. It is not stopping at the last line I have pasted above. I closed the batch file because it was taking a long time.

It is not generating in the same order as it did for you. it is generating in this order:
VC6, VC7, VC7.1, VC8, VC11 , VC12, VC14, VC15, VC11, VC12, VC14...

I ran the batch file from 7.74 release.

**Reply by jay** [reply]
This was a tricky one. Please try https://github.com/curl/curl/pull/6442

**Comment by ganeshkamath89** [comment]
Hi @jay. I like the solution given in Curl project where the batch file generates the VS projects. 

In my opinion, lot of people who use Curl also use OpenSSL. But OpenSSL does not provide any mechanism to debug the code by creating a VS project. Can you show your solution to the OpenSSL forum?

**Reply by jay** [reply]
> In my opinion, lot of people who use Curl also use OpenSSL. But OpenSSL does not provide any mechanism to debug the code by creating a VS project. Can you show your solution to the OpenSSL forum?

It's not transferable, their build process is too different. If your OpenSSL was built with debugging symbols then you can debug OpenSSL from the curl solution by stepping into OpenSSL functions or just open the file and put a breakpoint.

![Capture](https://user-images.githubusercontent.com/965580/104418972-c8820280-5545-11eb-92a2-ee941a41f63f.PNG)

**Reply by ganeshkamath89** [reply]
Thanks @jay  for sharing the solution. I am familiar with this technique. It is very cumbersome. 

I am very happy that Curl has a VS project support.

**Comment by jacobjake1** [comment]
So how do I do this in 2022? Generate.bat doesn't even exist??

**Reply by jay** [reply]
generate.bat is in the [projects](https://github.com/curl/curl/tree/master/projects) folder

**Reply by jacobjake1** [reply]
Im having an issue loading almost any solution file in VS, my current error is;

"C:\Users\jacob\AppData\Roaming\curlLoader.vcxproj : error  : Project "C:\Users\jacob\AppData\Roaming\curlLoader.vcxproj" could not be found."

None of these VC folders that it generated have a curlloader, how can I go about that?

**Reply by jay** [reply]
> curlLoader.vcxproj

That's not a project file from the curl project.
